client/block-builder/src/lib.rs

```diff
@@ -289,6 +289,15 @@ where
 			size
 		}
 	}
+
+	/// Execute any logic after processing all inherents in the block.
+	pub fn on_post_inherent(&self) -> Result<(), Error> {
+		let block_id = &self.block_id;
+
+		self.api
+			.on_post_inherent_with_context(&block_id, ExecutionContext::BlockConstruction)?;
```

Comment:
what will happen if runtime have not yet implemented this API?

Reply:
Block Production and Block Execution would come to different results.

Reply:
I think we need to think bit more about how to upgrade.

New runtime + old client needs to panic, else I think it will cause forking
Old runtime + new client should have the same behavior as old runtime + old client else it is forking

Reply:
syncing should be fine since there will be no implementations of `on_post_inherent` before this PR, so everything will be a noop.

I agree, this is clearly a client breaking change, and should be handled like any hard change, but also only if the feature is being used.

There is not much danger to release this in the wild as long as the feature is not used.

Reply:
if the client call the on_post_inherent api but the api doesn't exist in the runtime, I don't think it will be no-op. I think it will panic, no ?
there should be a way to only execute on_post_inherent if the api is provided by the runtime (by using version or stuff like this), this way the client support all kind of runtime.

Thought the client will have to upgrade before the runtime, an old client will produce invalid block.

Reply:
The more concrete solution to me would be:
* bump the version of BlockBuilder
* in the client we check the version of BlockBuilder and call into on_post_inherent only if version is superior or equal to the new version

When release a version we need to say that client must upgrade before runtime. or at least before runtime implement `on_post_inherent` with something not no-op.